### PR TITLE
Use helpers in param-nullchecking emit

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #36024 

This PR hit a snag related to the fact that the SynthesizedGlobalMethodSymbol has a `null` ContainingType, and that PrivateImplementationDetails is pretty emit-layer oriented. We could consider moving the lowering of param-nullchecking into the emit layer instead.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs [hunk 1]
```diff
@@ -577,7 +577,12 @@ public override BoundNode VisitLoweredConditionalAccess(BoundLoweredConditionalA
                 return null;
             }
 
-            if (method.ContainingType.IsAnonymousType)
+            if (method.ContainingType is null)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:580]: >method.ContainingType is null [](http://example.com/codeflow?start=16&length=29)

Is this case for the "Throw" methods? I'm curious why this wasn't needed for the "ComputeStringHash" method.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowMethod.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeGen;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// Throws a System.ArgumentNullException with the given 'paramName'.
+    /// </summary>
+    internal sealed class SynthesizedThrowMethod : SynthesizedGlobalMethodSymbol
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowMethod.cs:13]: >SynthesizedThrowMethod [](http://example.com/codeflow?start=26&length=22)

Is the separate "Throw" method necessary or could this be folded into "ThrowIfNull"?

File: src/Compilers/CSharp/Portable/Symbols/MemberSymbolExtensions.cs [hunk 3]
```diff
@@ -453,7 +454,7 @@ internal static bool IsParameterlessConstructor(this MethodSymbol method)
         internal static bool IsDefaultValueTypeConstructor(this MethodSymbol method, bool requireZeroInit)
         {
             if (method.IsImplicitlyDeclared &&
-                method.ContainingType.IsValueType &&
+                method.ContainingType?.IsValueType == true &&
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/MemberSymbolExtensions.cs:457]: >method.ContainingType?.IsValueType == true [](http://example.com/codeflow?start=16&length=42)

Any idea why this change wasn't needed for "ComputeStringHash" previously?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowIfNullMethod.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            SyntheticBoundNodeFactory F = new SyntheticBoundNodeFactory(this, this.GetNonNullSyntaxNode(), compilationState, diagnostics);
+            F.CurrentFunction = this;
+
+            try
+            {
+                ParameterSymbol argument = this.Parameters[0];
+                ParameterSymbol paramName = this.Parameters[1];
+
+                MethodSymbol throwMethod = _throwMethod;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowIfNullMethod.cs:36]: >throwMethod [](http://example.com/codeflow?start=29&length=11)

Consider inlining since there's only use.

File: src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs [hunk 5]
```diff
@@ -1456,14 +1525,63 @@ class C
...
+  .maxstack  2
+  IL_0000:  ldarg.1
+  IL_0001:  ldstr      ""number""
+  IL_0006:  call       ""ThrowIfNull""
+  IL_000b:  ret
+}");
+        }
+
+        [Fact]
+        public void TestTypeParameter()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs:1538]: >TestTypeParameter [](http://example.com/codeflow?start=20&length=17)

Consider testing type parameters with constraints as well: { `struct`, `class`, `notnull` }.

File: src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:580]: It's because in that case, we never include the method symbol in the bound tree, and simply grab the CCI adapter for the symbol and use it directly in the emit layer￼.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowMethod.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowMethod.cs:13]: I was instructed to factor it in precisely this way. Will make sure to include the reference to the similar methods on ArgumentNullException in the runtime.￼￼

File: src/Compilers/CSharp/Portable/Symbols/MemberSymbolExtensions.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/MemberSymbolExtensions.cs:457]: This is the first time we are using "synthesized global methods" in the bound tree.￼

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 6]
```diff
@@ -1777,6 +1779,35 @@ internal void EnsureNativeIntegerAttributeExists()
             EnsureEmbeddableAttributeExists(EmbeddableAttributes.NativeIntegerAttribute);
         }
 
+#nullable enable
+        /// <summary>
+        /// Creates the ThrowIfNull and Throw helpers if needed.
+        /// </summary>
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:1785]:
```suggestion
        /// <summary>
        /// Creates the ThrowIfNull and Throw helpers if needed.
        /// </summary>
        /// <remarks>
        /// The ThrowIfNull and Throw helpers are modeled off of the helpers on ArgumentNullException.
        /// https://github.com/dotnet/runtime/blob/22663769611ba89cd92d14cfcb76e287f8af2335/src/libraries/System.Private.CoreLib/src/System/ArgumentNullException.cs#L56-L69
        /// </remarks
```

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 7]
```diff
@@ -1777,6 +1779,39 @@ internal void EnsureNativeIntegerAttributeExists()
...
+        /// <remarks>
+        /// The ThrowIfNull and Throw helpers are modeled off of the helpers on ArgumentNullException.
+        /// https://github.com/dotnet/runtime/blob/22663769611ba89cd92d14cfcb76e287f8af2335/src/libraries/System.Private.CoreLib/src/System/ArgumentNullException.cs#L56-L69
+        /// </remarks
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:1789]:
```suggestion
        /// </remarks>
```

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowMethod.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowMethod.cs:13]: Done. See recent changes in PEModuleBuilder.cs.

**Review by cston — APPROVED** [review]

**Comment by RikkiGibson** [comment]
I think integration test failures are due to need to ingest newer changes from main into the feature branch.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 8]
```diff
@@ -1777,6 +1779,39 @@ internal void EnsureNativeIntegerAttributeExists()
...
+            TypeSymbol returnType = factory.SpecialType(SpecialType.System_Void);
+            TypeSymbol argumentType = factory.SpecialType(SpecialType.System_Object);
+            TypeSymbol paramNameType = factory.SpecialType(SpecialType.System_String);
+
+            var sourceModule = SourceModule;
+            var throwMethod = new SynthesizedThrowMethod(sourceModule, privateImplClass, returnType, paramNameType);
+            privateImplClass.TryAddSynthesizedMethod(throwMethod.GetCciAdapter());
+
+            var throwIfNullMethod = new SynthesizedThrowIfNullMethod(sourceModule, privateImplClass, throwMethod, returnType, argumentType, paramNameType);
+            privateImplClass.TryAddSynthesizedMethod(throwIfNullMethod.GetCciAdapter());
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:1810]: It feels like, if this returns false, we should be getting the final throwIfNull method that some other thread succeeded in adding, rather than returning the symbol we failed to add. It doesn't look like any use of `TryAddSynthesizedMethod` today is used by a method that returns a value, so this would be a new change.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs [hunk 9]
```diff
@@ -49,28 +62,16 @@ internal sealed partial class LocalRewriter
...
-
-            var argumentName = ImmutableArray.Create<BoundExpression>(factory.StringLiteral(parameter.Name));
-            BoundObjectCreationExpression ex = factory.New(factory.WellKnownMethod(WellKnownMember.System_ArgumentNullException__ctorString), argumentName);
-            BoundThrowStatement throwArgNullStatement = factory.Throw(ex);
-
-            return factory.HiddenSequencePoint(factory.If(paramIsNullCondition, throwArgNullStatement));
+            var call = factory.Call(
+                receiver: null,
+                throwIfNullMethod,
+                arg0: factory.Convert(factory.SpecialType(SpecialType.System_Object), factory.Parameter(parameter)),
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:72]: Do we have tests for `void M(void* ptrParam!!)`?

Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:72]: This will box nullable value types. That doesn't feel like an optimization.

**Comment by 333fred** [comment]
Done review pass (commit 5). I did not look at test changes.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:1810]: Ahh, you're right. It feels like if we fail to add the first synthesized method here, we need to spin-wait or something until we get the second synthesized method that we expect another thread to add.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs (hunk 9 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:72]: We don't.. but it is an error in the current implementation, and the [current proposal](https://github.com/dotnet/csharplang/blob/3a6c79012a8c53fe84330c09080105ac177ff93d/proposals/null-arg-checking.md) doesn't mention it.

It seems not-wrong to me to support pointers with the feature. We did after all even make `ptr is null` legal at some point. Perhaps an email to LDT would be reasonable and it's more motivation to restore the "non-helper" code path.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:72]: This is true. However we considered nullable value types a pathological scenario to the point where I judged keeping the previous code path around for it was not worthwhile. This is also why we give a warning in this scenario.

Given the similar concerns about pointers (preferable to use "existing" codegen) perhaps we may want to restore the behavior. Or generate and use only the "Throw" helper in this case and reserve the "ThrowIfNull" helper for things which are known to be reference types.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 8 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:1810]: I wouldn't bother spin waiting on the first one, we'll do a bit of extra work, but not much. Instead just gate it on the second add, and re-do the lookup if it fails. (You may want to dig into how much work creating them actually is compared with the lookup and just always try and add them and only return the already created one if it fails)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowMethod.cs (hunk 2 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowMethod.cs:13]: I think(?) it's to do with inlining. IIRC the JIT wont inline a method with a `throw` in it. So `ThrowIfNull` now gets inlined on the fast path, and the overhead of the extra method call to `Throw` is negated by the fact that we only call it when we already know we're throwing an exception.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs (hunk 9 above)
Comment (chsienki) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:72]: It does seem odd. I can try and convince myself of scenarios, but why would I write a method that accepts a nullable value type, only to throw if it's null. You'd just make it take the value type and force the caller to handle it being nullable. Like, we have a compile time check for this, when would I ever want to fallback to a runtime check instead?

File: src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs]: Done.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: I went ahead and made this change, although after thinking about it a bit I suspect that the specific instance of the method symbol between the PrivateImplementationDetails class and a particular call site of the helper method is not going to end up affecting compiler behavior.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: It likely won't, until it ends up being exposed via some public API and then all hell will break loose.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 10]
```diff
@@ -1777,6 +1779,60 @@ internal void EnsureNativeIntegerAttributeExists()
...
+            TypeSymbol argumentType = factory.SpecialType(SpecialType.System_Object);
+            TypeSymbol paramNameType = factory.SpecialType(SpecialType.System_String);
+
+            var sourceModule = SourceModule;
+            var throwMethod = new SynthesizedThrowMethod(sourceModule, privateImplClass, returnType, paramNameType);
+            privateImplClass.TryAddSynthesizedMethod(throwMethod.GetCciAdapter());
+
+            // If this call lost a race with another call, we want to obtain and return the symbol that was actually added successfully.
+            // note: strictly it is possible for another thread to add the 'throwMethod', then suspend, then this thread adds the 'throwIfNullMethod'.
+            // If this situation occurs it is not really a concern. We will emit all the methods we need and even if we use a different object instance
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:1810]: >If this situation occurs it is not really a concern. [](http://example.com/codeflow?start=15&length=52)

> If this situation occurs it is not really a concern.

I don't agree. If we ever store these things in a dictionary, or ever expose them through a public API, we have a nasty bug waiting for us. Let's make sure that we're not accidentally introducing bad race conditions here.

**Comment by 333fred** [comment]
Done review pass (commit 6). Have not looked at test changes.

**Comment by RikkiGibson** [comment]
@333fred I made the changes to restore the old nullable value type behavior, support pointers and function pointers, and fix adding the synthesized methods to PrivateImplementationDetails.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs [hunk 11]
```diff
@@ -49,10 +53,28 @@ internal sealed partial class LocalRewriter
...
+                arg1: factory.StringLiteral(parameter.Name));
+            return factory.HiddenSequencePoint(factory.ExpressionStatement(call));
         }
 
-        private static BoundStatement ConstructIfStatementForParameter(ParameterSymbol parameter, SyntheticBoundNodeFactory factory)
+        private static BoundStatement ConstructDirectNullCheck(ParameterSymbol parameter, SyntheticBoundNodeFactory factory)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:77]: Consider asserting the type of the parameter here.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:77]: Oh, actually, you check it down below. That's fine.

**Comment by 333fred** [comment]
Test failures look legitimate. Done review pass (commit 9)

**Comment by RikkiGibson** [comment]
Bleh, yes, basically when a type parameter is constrained to some flavor of `System.Nullable<T>` it's a bit of a pain to actually access the HasValue member on it. I added 2 commits to try and resolve.

The first commit e40037f actually attempts to access the HasValue member on the type parameter constrained to nullable value type. This introduced a verification failure that I wasn't expecting: ``PeVerify failed for assembly '':
[ : B3`1[T]::M[U]][mdToken=0x6000003][offset 0x00000002][found address of ref ][expected address of value 'System.Nullable`1[T]'] Unexpected type on the stack.``. I wonder if a separate bug in lowering is causing us to emit a `ldarga` when we should be emitting a `ldarg`.

It seems like the code actually executes just fine at least on core. Still it made me a bit wary and since the scenario is so out in left field I pushed 00a2898 which reverts most of its parent and uses the "boxing" code gen in this specific case.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs (hunk 9 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:72]: Fixed by adding support for pointer types.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: Fixed by using a lock.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 12]
```diff
@@ -1777,6 +1779,44 @@ internal void EnsureNativeIntegerAttributeExists()
...
+        /// Creates the ThrowIfNull and Throw helpers if needed.
+        /// </summary>
+        /// <remarks>
+        /// The ThrowIfNull and Throw helpers are modeled off of the helpers on ArgumentNullException.
+        /// https://github.com/dotnet/runtime/blob/22663769611ba89cd92d14cfcb76e287f8af2335/src/libraries/System.Private.CoreLib/src/System/ArgumentNullException.cs#L56-L69
+        /// </remarks>
+        internal MethodSymbol EnsureThrowIfNullFunctionExists(SyntaxNode syntaxNode, SyntheticBoundNodeFactory factory, DiagnosticBag? diagnostics)
+        {
+            var privateImplClass = GetPrivateImplClass(syntaxNode, diagnostics);
+            lock (privateImplClass)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:1793]: >lock (privateImplClass) [](http://example.com/codeflow?start=12&length=23)

Could we avoid the lock by always using the symbols from `PrivateImplementationDetails`?
```csharp
if (privateImplClass.GetMethod("ThrowIfNull") is null)
{
    privateImplClass.TryAddSynthesizedMethod(new SynthesizedThrowMethod(...));
    privateImplClass.TryAddSynthesizedMethod(
        new SynthesizedThrowIfNullMethod(...
            privateImplClass.GetMethod("Throw"));
}
return privateImplClass.GetMethod("ThrowIfNull");
```

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:1793]: I guess this does guarantee we would use the ThrowMethod that really got added :)

**Review by 333fred — APPROVED** [review]
LGTM (commit 12)

**Comment by cston** [comment]
>             return factory.HiddenSequencePoint(factory.If(paramIsNullCondition, throwArgNullStatement));

I'm curious why we chose to inline the `if (arg is null) throw new ArgumentNullException();` in these cases rather than using a helper.

---
Refers to: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullChecking.cs:102 in c345bd5. [](commit_id = c345bd58eb4cd1248adbea44a6a70804d43e21b5, deletion_comment = False)

File: src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs [hunk 13]
```diff
@@ -1456,14 +1547,189 @@ class C
...
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  brtrue.s   IL_000e
+  IL_0003:  ldstr      ""funcPtr""
+  IL_0008:  newobj     ""System.ArgumentNullException..ctor(string)""
+  IL_000d:  throw
+  IL_000e:  nop
+  IL_000f:  ret
 }");
         }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs:1734]: >} [](http://example.com/codeflow?start=8&length=1)

Consider verifying that we did not generate `ThrowIfNull()` and `Throw()` methods in these last three tests that inline the checks.
```csharp
verifier.VerifyMissing("<PrivateImplementationDetails>.ThrowIfNull");
verifier.VerifyMissing("<PrivateImplementationDetails>.Throw");
```

**Comment by RikkiGibson** [comment]
> I'm curious why we chose to inline the if (arg is null) throw new ArgumentNullException(); in these cases rather than using a helper.

For pointers it didn't feel worthwhile to come up with another pair of helpers (ThrowPointerIfNull, etc.) due to their uncommonness. For nullable value types it seemed like it would require a generic `ThrowIfNull<T>(T? argument, string paramName)` which also doesn't seem worth it.

FWIW, I imagine we may evolve this implementation in the future. If the runtime changes to inline methods across assemblies in more scenarios, we might be able to drop the internal helpers and use directly the `ArgumentNullException.ThrowIfNull` helpers defined for both objects and pointers.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs (hunk 13 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenNullCheckedParameterTests.cs:1740]: Done

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowIfNullMethod.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedThrowIfNullMethod.cs]: Fixed.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: Well, fixed by adding-then-getting, accomplishing the same thing.

**Comment by RikkiGibson** [comment]
roslyn-integration-ci won't pass due to being out of date with main. Merging now.
